src/python/pants/goal/run_tracker.py

```diff
@@ -433,13 +436,19 @@ def store_stats(self):
 
   _log_levels = [Report.ERROR, Report.ERROR, Report.WARN, Report.INFO, Report.INFO]
 
+  def has_ended(self):
+    return self._end_memoized_result is not None
+
+  # TODO: memoize this with a decorator!
   def end(self):
```

Comment:
Why do we now memoize this? This seems like a thing that should only be called once?

Reply:
The docstring says subsequent calls are no-ops, so I didn't think it would break the existing contract to explicitly memoize it. It may not be necessary to do that now given the other changes that have happened since this part was written. Also, using a `@memoized_method` decorator should probably be done, I recall it happened to be failing several months ago when I did that but don't recall exactly why.

Reply:
It feels like either this is necessary, or this code https://github.com/pantsbuild/pants/pull/6574/files#diff-3fdeeecea726b1ee5108408dbaf23ae1R60 which handles errors for running end twice would be necessary, but not both? Am I mis-understanding what the `ValueError` catching there is meant to handle?

Reply:
> It may not be necessary to do that now given the other changes that have happened since this part was written.

No, you are understanding it correctly!

Reply:
Ok, reviewing it again. The reason we have changes in this file are to allow the `.has_ended()` call in that file, which is called in `local_pants_runner.py`'s `LocalExiter`. The `ValueError` being caught is for catching a write to a closed file, which can occur if `.end()` hasn't been called yet, but `.exit()` is getting called in a signal handler. It seems like I believe you are suggesting that it might make sense to just call `.end()` and handle whatever error occurs from calling that in an `except`.